sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala

```diff
@@ -200,14 +200,20 @@ object ConstantPropagation extends Rule[LogicalPlan] {
 
   private def replaceConstants(condition: Expression, equalityPredicates: EqualityPredicates)
     : Expression = {
-    val constantsMap = AttributeMap(equalityPredicates.map(_._1))
-    val predicates = equalityPredicates.map(_._2).toSet
-    def replaceConstants0(expression: Expression) = expression transform {
+    val allConstantsMap = AttributeMap(equalityPredicates.map(_._1))
+    val allPredicates = equalityPredicates.map(_._2).toSet
+    def replaceConstants0(
+        expression: Expression, constantsMap: AttributeMap[Literal]) = expression transform {
       case a: AttributeReference => constantsMap.getOrElse(a, a)
     }
     condition transform {
-      case e @ EqualTo(_, _) if !predicates.contains(e) => replaceConstants0(e)
-      case e @ EqualNullSafe(_, _) if !predicates.contains(e) => replaceConstants0(e)
+      case b: BinaryComparison =>
+        if (!allPredicates.contains(b)) {
+          replaceConstants0(b, allConstantsMap)
+        } else {
+          val excludedEqualityPredicates = equalityPredicates.filterNot(_._2.semanticEquals(b))
```

Comment:
Exclude current binary comparison to support the following case:
```
a = 1 and a = 2 ==> 2 = 1 and 1 = 2
```

Reply:
this doesn't look like constant propagation but a different optimization to simplify predicates:
- a = 1 and a = 2 ==> false
- a > 1 and a > 2 ==> a > 2

Reply:
This PR does not support `a > 1 and a > 2 ==> a > 2`. It must contain an equality predicate.

Reply:
Can we exclude it first? We should add a new optimizer rule later to do more advanced predicate simplification.

Reply:
Do you mean only support `a = 1 and a = 2 ==> 2 = 1 and 1 = 2 ==> false`, and do not support `a = 1 and b > a + 2 ==> a = 1 && b > 3`?

Reply:
nvm, the rule name is `ConstantPropagation`, so both optimizations fit it.